### PR TITLE
Refactor link creation

### PR DESCRIPTION
**Pull request by kwk**
Once https://github.com/almighty/almighty-core/pull/1444 is merged, we need to adopt the UI as well.

This is my attempt to do so but I haven't tested it, just went over the code and did the follwing:

1. Remove the source and target type fields from the work item link type
2. Don't filter by type when searching doing the type-ahead.

I would love to get some feedback on this because most of what I did was assisted guessing ;)

### CODE REVIEW

**Comment by fabric8cd** [comment]
@kwk snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2042-fabric8-planner.int.rdu2c.fabric8.io

File: src/app/services/work-item.service.ts [hunk 1]
```diff
@@ -946,12 +938,12 @@ export class WorkItemService {
...
   }
 
-  searchLinkWorkItem(term: string, workItemType: string): Observable<WorkItem[]> {
+  searchLinkWorkItem(term: string): Observable<WorkItem[]> {
     if (this._currentSpace) {
       // FIXME: make the URL great again (when we know the right API URL for this)!
       // search within selected space
-      let searchUrl = this.baseApiUrl + 'search?spaceID=' + this._currentSpace.id + '&q=' + term + ' type:' + workItemType;
-      //let searchUrl = currentSpace.links.self + 'search?q=' + term + ' type:' + workItemType;
+      let searchUrl = this.baseApiUrl + 'search?spaceID=' + this._currentSpace.id + '&q=' + term;
```
Comment (sanbornsen) [inline src/app/services/work-item.service.ts]: You are not passing the type anymore, does that mean it fetches all the work items base on the keywords?

File: src/app/services/work-item.service.ts [hunk 2]
```diff
@@ -779,16 +773,14 @@ export class WorkItemService {
...
-        wiType: linkType.relationships.target_type.data.id
+        linkType: 'forward'
       });
     });
     linkTypes.backwardLinks.forEach((linkType: LinkType) => {
       opLinkTypes.push({
         name: linkType.attributes['reverse_name'],
         linkId: linkType.id,
-        linkType: 'reverse',
-        wiType: linkType.relationships.source_type.data.id
```
Comment (sanbornsen) [inline src/app/services/work-item.service.ts]: Why did you remove wiType ?

File: src/app/services/work-item.service.ts [hunk 3]
```diff
@@ -731,46 +731,40 @@ export class WorkItemService {
...
-      return linkTypes;
-    })
-    .map((linkTypes: any) => {
-      return this.formatLinkTypes(linkTypes);
-    })
-    .catch((err) => {
-      console.log(err);
-      return Observable.of({});
-    });
+    return Observable.forkJoin(this.getAllLinkTypes(workItem))
```
Comment (sanbornsen) [inline src/app/services/work-item.service.ts]: You don;t have to use forkJoin in that case. You can replace the line with 
```return this.getAllLinkTypes(workItem)```

File: src/app/services/work-item.service.ts [hunk 4]
```diff
@@ -731,46 +731,40 @@ export class WorkItemService {
...
+        .map(items => {
+          let linkTypes: Object = {};
+          linkTypes['forwardLinks'] = items[0].json().data;
+          linkTypes['backwardLinks'] = items[1].json().data;
```
Comment (sanbornsen) [inline src/app/services/work-item.service.ts]: Now items is not an array of array, so it'll fail here. 
Items would be one array having all the link types.
How would you separate forward and backward link here?

File: src/app/services/work-item.service.ts (hunk 3 above)
Comment (kwk) [inline src/app/services/work-item.service.ts]: I thought so too but I didn't want to change much of the logic.

File: src/app/services/work-item.service.ts (hunk 2 above)
Comment (kwk) [inline src/app/services/work-item.service.ts]: Because it no longer exists, since the `source_type` field has been removed from the link type.

File: src/app/services/work-item.service.ts (hunk 1 above)
Comment (kwk) [inline src/app/services/work-item.service.ts]: yes and that is intentional

File: src/app/services/work-item.service.ts (hunk 4 above)
Comment (kwk) [inline src/app/services/work-item.service.ts]: Having another look.

File: src/app/services/work-item.service.ts (hunk 1 above)
Comment (sanbornsen) [inline src/app/services/work-item.service.ts]: What if the work item type doesn't match with the link type?

Comment (kwk) [inline src/app/services/work-item.service.ts]: > What if the work item type doesn't match with the link type?

We have no restrictions on the backend any longer once the PR in the description is merged.

File: src/app/services/work-item.service.ts (hunk 4 above)
Comment (kwk) [inline src/app/services/work-item.service.ts]: @sanbornsen fixed in 63cbd8f

File: src/app/mock/mock-http.ts [hunk 5]
```diff
@@ -247,10 +247,6 @@ export class MockHttp extends HttpService {
           } else {
             return this.createResponse(url.toString(), 200, 'ok', { data: this.mockDataService.getAllAreas() });
           }
-        case '/source-link-types':
```
Comment (sanbornsen) [inline src/app/mock/mock-http.ts]: We will need **/workitemlinktypes** mock data service. Maybe just keep one of these and replace **/target-link-types** with **/workitemlinktypes**

**Comment by fabric8cd** [comment]
@kwk snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2042-fabric8-planner.int.rdu2c.fabric8.io

File: src/app/mock/mock-http.ts (hunk 5 above)
Comment (kwk) [inline src/app/mock/mock-http.ts]: okay

Comment (kwk) [inline src/app/mock/mock-http.ts]: I've added @sanbornsen I've added the mock endpoint in 	8eba828 and also removed a link category that was included in the link type's `included` array but doesn't belong to the link type in 9f9aff8.

Comment (sanbornsen) [inline src/app/mock/mock-http.ts]: Cool!! all green.

**Comment by fabric8cd** [comment]
@kwk snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2042-fabric8-planner.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@kwk snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-2042-fabric8-planner.int.rdu2c.fabric8.io
